lib/raty.js

```diff
@@ -0,0 +1,784 @@
+    this._adjustHints();
+
+    this.opt.score = this._adjustedScore(this.opt.score);
+
+    if (this.opt.starType !== 'img') {
+      this._adjustStarName();
+    }
+
+    this._setPath();
+    this._createStars();
+
+    if (this.opt.cancelButton) {
+      this._createCancel();
+    }
+
+    if (this.opt.precision) {
+      this._adjustPrecision();
+    }
+
+    this._createScore();
+
+    this._apply(this.opt.score);
+    this._setTitle(this.opt.score);
+    this._target(this.opt.score);
+
+    if (this.opt.readOnly) {
+      this._lock();
+    } else {
+      this.self.style.cursor = 'pointer';
+      this._binds();
+    }
+  }
+
+  //helper functions
+  // TODO: model spec
+  _adjustedScore(score) {
+    if (score || score === 0) {
+      return this._between(score, 0, this.opt.number);
+    }
+  }
+
+  _adjustHints() {
+    if (!this.opt.hints) {
+      this.opt.hints = [];
+    }
+
+    if (!this.opt.halfShow && !this.opt.half) {
+      return;
+    }
+
+    var steps = this.opt.precision ? 10 : 2;
+
+    for (var i = 0; i < this.opt.number; i++) {
+      var group = this.opt.hints[i];
+
+      if (Object.prototype.toString.call(group) !== '[object Array]') {
+        group = [group];
+      }
+
+      this.opt.hints[i] = [];
+
+      for (var j = 0; j < steps; j++) {
+        var hint = group[j];
+        var last = group[group.length - 1];
+
+        if (last === undefined) {
+          last = null;
+        }
+
+        this.opt.hints[i][j] = hint === undefined ? last : hint;
+      }
+    }
+  }
+
+  _adjustNumber() {
+    this.opt.number = this._between(this.opt.number, 1, this.opt.numberMax);
+  }
+
+  _adjustPrecision() {
+    this.opt.half = true;
+  }
+
+  _adjustStarName() {
+    var replaces = ['cancelOff', 'cancelOn', 'starHalf', 'starOff', 'starOn'];
+
+    this.opt.path = '';
+
+    for (var i = 0; i < replaces.length; i++) {
+      this.opt[replaces[i]] = this.opt[replaces[i]].replace('.', '-');
+    }
+  }
+
+  // TODO: model spec
+  _apply(score) {
+    this._fill(score);
+
+    if (score) {
+      if (score > 0) {
+        this.scoreField.value = score;
+      }
+
+      this._roundStars(score);
+    }
+  }
+
+  _attributesForIndex(i) {
+    var name = this._nameForIndex(i);
+    var attributes = { alt: i, src: this.opt.path + this.opt[name] };
+
+    if (this.opt.starType !== 'img') {
+      attributes = { 'data-alt': i, 'class': this.opt[name] };
+    }
+
+    attributes.title = this._getHint(i);
+
+    return attributes;
+  }
+
+  _between(value, min, max) {
+    return Math.min(Math.max(parseFloat(value), min), max);
+  }
+
+  // TODO: model spec
+  _binds() {
+    if (this.cancelButton) {
+      this._bindOverCancel();
+      this._bindClickCancel();
+      this._bindOutCancel();
+    }
+    this._bindOver();
+    this._bindClick();
+    this._bindOut();
+  }
+
+  // TODO: model spec
+  _bindClick() {
+    let that = this;
+    this.stars.forEach((value) => {
+      value.addEventListener('click', function (evt) {
```

Comment:
Before we were using the `click.raty` or any action name plus `.raty`, so we could do `.off('.raty')` what would remove all listeners for that scope. I don't know if we have the equivalent, but if not where we use the off we should understand which listener should be turned off if not all.

Another bad thing about it is that we can remove an external binding that does not belong to us, so that's why I set it. But maybe it's not a problem, but let's keep eyes on it.

Reply:
That's why I used the pointer-event CSS property.

[see the conversation](https://github.com/wbotelhos/raty/pull/240#discussion_r980737775)